src/index.ts

```diff
@@ -874,8 +875,16 @@ class ThreadPool {
   }
 
   _maybeDrain () {
-    if (this.taskQueue.size === 0 && this.skipQueue.length === 0) {
+    const totalCapacity = this.options.maxQueue + this.pendingCapacity();
+    const totalQueueSize = this.taskQueue.size + this.skipQueue.length;
+
+    if (totalQueueSize === 0) {
       this.publicInterface.emit('drain');
+      this.needsDrain = false;
```

Comment:
why not `this.pool.needsDrain`?

Reply:
Do you mean to hold it at the `publicInterface` instead?

Reply:
I think the needsDrain = false should be executed when ` (totalQueueSize >= totalCapacity) == false`. If you reset only when is 0, the queue will not be full, and `needsDrain` will show true.

Reply:
Please check how writable does. I think it sets needsSrain to false before emitting drain.